feat: updates for cbor all the things (#955)
datatrails/datatrails-docs: content/developers/developer-patterns/scitt-api/index.md

    ISSUER="sample.synsation.io"

    # signing key to sign the SCITT Statements
    SIGNING_KEY="my-signing-key.pem"

    # File representing the signed statement to be registered
    SIGNED_STATEMENT_FILE="signed-statement.cbor"

    # File representing the transparent statement, which includes the signed statement and the registration receipt
    TRANSPARENT_STATEMENT_FILE="transparent-statement.cbor"

    # Property used to correlate a collection of statements about an artifact
    SUBJECT="my-product-id"

> 

Author: Using `python -m` as it works universaly on linux, mac and windows

    ISSUER="sample.synsation.io"

    # signing key to sign the SCITT Statements

>     SIGNING_KEY="my-signing-key.pem"

Reviewer: Is there a danger that this will create secrets in the git repo that may then be accidentally pushed to github?

Reviewer: we usually use this construct which is Mac compatible I think:

```
TEMPDIR=$(mktemp -d)
function finalise {
    CODE=$?
    rm -rf "$TEMPDIR"
    exit $CODE
}
trap finalise EXIT
```

Author: `$HOME` aslo doesn't work on macos :-(

Author: I used this idiom for curl auth headers. it is like our current pattern with a tiny tweak so the header can be provided directly or indirectily via file

```
# The authentication header for curl commands, which *may* be a file ref if
# prefixed with@
BEARER_TOKEN_HEADER=@~/.datatrails/bearer-token.txt
```

But the key detail is that the OS sensitive bits need to be configurable and not baked into the examples themselves

Author: I've added a note regarding changing the config to refer to the OS temp location

    ISSUER="sample.synsation.io"

    # signing key to sign the SCITT Statements
    SIGNING_KEY="my-signing-key.pem"

    # File representing the signed statement to be registered
    SIGNED_STATEMENT_FILE="signed-statement.cbor"

    # File representing the transparent statement, which includes the signed statement and the registration receipt
    TRANSPARENT_STATEMENT_FILE="transparent-statement.cbor"

    # Property used to correlate a collection of statements about an artifact
    SUBJECT="my-product-id"

>     # A command which produces a hash, eg sha256sum on linux, or shasum on macos

Reviewer 2: ```suggestion
    # A command which produces a hash, eg sha256sum on linux, or shasum on Mac OS
```

>     # The specific algorithm is not important for these examples

Reviewer 2: ```suggestion
    # SHA-256 is used here, but alternative algorithms are also supported
```

    HASH_COMMAND=sha256sum
    ```

{{< note >}}

> These defaults will place files in your current working directory. For session persistence, consider replacing the file paths with absolute paths. For example `SIGNING_KEY="$HOME/.datatrails/my-signing-key.pem"`

Reviewer 2: For these samples, I'd like to keep things super simple, avoiding lots of discussion a developer should be able to infer. 
```suggestion
These defaults will place files in your current working directory.
For session persistence, consider replacing the file paths with absolute paths.
For example `SIGNING_KEY="$HOME/.datatrails/my-signing-key.pem"`
```

{{< /note >}}

> 

Reviewer 2: ```suggestion
```

    ISSUER="sample.synsation.io"

    # signing key to sign the SCITT Statements
    SIGNING_KEY="my-signing-key.pem"

    # File representing the signed statement to be registered
    SIGNED_STATEMENT_FILE="signed-statement.cbor"

    # File representing the transparent statement, which includes the signed statement and the registration receipt
    TRANSPARENT_STATEMENT_FILE="transparent-statement.cbor"

    # Property used to correlate a collection of statements about an artifact
    SUBJECT="my-product-id"

>     # A command which produces a hash, eg sha256sum on linux, or shasum on macos
>     # The specific algorithm is not important for these examples
>     HASH_COMMAND=sha256sum

Reviewer 2: Do we expect users to change this? The sample uses the same command on all platforms. Is there a reason to keep this as an envvar?
```suggestion
```

## Create Metadata

[DataTrails Event Attributes](./../../api-reference/events-api/) can be associated with a SCITT Statement, enabling indexing and retrieval in future releases.

Create metadata with a dictionary of `key:value` pairs.

```bash

> HASH=$($HASH_COMMAND "payload.json" | cut -d ' ' -f 1)

Reviewer 2: See comment at: https://github.com/datatrails/datatrails-docs/pull/955/files#r1880440076
```suggestion
HASH=$(sha256sum "payload.json" | cut -d ' ' -f 1)
```

> python ${SCRIPTS}create_signed_statement.py \

Reviewer: its is usually python3.
on my laptop the cmd 'python' does not exist

Author: ah that is what I had originaly, but I think someone changed it to python3. happy to change it back

Author: python3 -m used consistently now (I hope)

1. Submit the Signed Statement to DataTrails, using the credentials in the `DATATRAILS_CLIENT_ID` and `DATATRAILS_CLIENT_SECRET`.

    ```bash
    python -m datatrails_scitt_samples.scripts.register_signed_statement \
      --signed-statement-file signed-statement.cbor \
      --output-file $TRANSPARENT_STATEMENT_FILE \
      --log-level INFO
    ```

    Find and copy the leaf hash from the output. It will look like this:
    ```
    INFO:register-statement:Leaf Hash: 30f5650fbe3355ca892094a3fbe88e5fa3a9ae47fe3d0bbace348181eb2b76db
    ```

1. View the Transparent Statement, as a result of registering the Signed Statement

    ```bash
    python -m datatrails_scitt_samples.dump_cbor \
      --input transparent-statement.cbor
    ```

1. Verify the the receipt

    ```bash

>     python -m datatrails_scitt_samples.scripts.verify_receipt \

Author: There is a companion PR for datatrails_scitt_samples that needs to merge in order for this to work

1. Submit the Signed Statement to DataTrails, using the credentials in the `DATATRAILS_CLIENT_ID` and `DATATRAILS_CLIENT_SECRET`.

    ```bash
    python -m datatrails_scitt_samples.scripts.register_signed_statement \

>       --signed-statement-file signed-statement.cbor \

Author: unintentional change thanks

1. Submit the Signed Statement to DataTrails, using the credentials in the `DATATRAILS_CLIENT_ID` and `DATATRAILS_CLIENT_SECRET`.

    ```bash
    python3 -m datatrails_scitt_samples.scripts.register_signed_statement \
      --signed-statement-file $SIGNED_STATEMENT_FILE \
      --output-file $TRANSPARENT_STATEMENT_FILE \
      --log-level INFO
    ```

    The last line of the output will include the leaf entry that commits the statement to the merkle log.
    It will look like
    ```
    {"entryid": "assets_b9d32c32-8ab3-4b59-8de8-bd6393167450_events_7dd2a825-495e-4fc9-b572-5872a268c8a9",
     "leaf": "30f5650fbe3355ca892094a3fbe88e5fa3a9ae47fe3d0bbace348181eb2b76db"}
    ```

    Add the `--log-level DEBUG` flag to help diagnose any issues.

1. View the Transparent Statement, as a result of registering the Signed Statement

    ```bash
    python3 -m datatrails_scitt_samples.dump_cbor \
      --input $TRANSPARENT_STATEMENT_FILE
    ```

1. Verify the the receipt

    ```bash
    python3 -m datatrails_scitt_samples.scripts.verify_receipt \
      --transparent-statement-file $TRANSPARENT_STATEMENT_FILE \
      --leaf $LEAF
    ```

>     Following the example above $LEAF should be:
>     `30f5650fbe3355ca892094a3fbe88e5fa3a9ae47fe3d0bbace348181eb2b76db`

Reviewer 2: ```suggestion
    Following the example above, $LEAF should be:

    ```output
    30f5650fbe3355ca892094a3fbe88e5fa3a9ae47fe3d0bbace348181eb2b76db
    ```

```

1. Submit the Signed Statement to DataTrails, using the credentials in the `DATATRAILS_CLIENT_ID` and `DATATRAILS_CLIENT_SECRET`.

    ```bash
    python3 -m datatrails_scitt_samples.scripts.register_signed_statement \
      --signed-statement-file $SIGNED_STATEMENT_FILE \
      --output-file $TRANSPARENT_STATEMENT_FILE \
      --log-level INFO
    ```

    The last line of the output will include the leaf entry that commits the statement to the merkle log.
    It will look like
    ```

>     {"entryid": "assets_b9d32c32-8ab3-4b59-8de8-bd6393167450_events_7dd2a825-495e-4fc9-b572-5872a268c8a9",
>      "leaf": "30f5650fbe3355ca892094a3fbe88e5fa3a9ae47fe3d0bbace348181eb2b76db"}

Reviewer 2: ```suggestion
    {
      "entryid": "assets_b9d32c32-8ab3-4b59-8de8-bd6393167450_events_7dd2a825-495e-4fc9-b572-5872a268c8a9",
      "leaf": "30f5650fbe3355ca892094a3fbe88e5fa3a9ae47fe3d0bbace348181eb2b76db"
     }
```

    ```bash
    python3 -m datatrails_scitt_samples.scripts.verify_receipt \
      --transparent-statement-file $TRANSPARENT_STATEMENT_FILE \
      --leaf $(jq -r .leaf <<<"$RESPONSE")
    ```

    The verification should pass with:

    ```output
    verification passed
    ```

1. Simulate a failed verification, by altering the `.leaf` value

    ```bash

    python3 -m datatrails_scitt_samples.scripts.verify_receipt \
      --transparent-statement-file $TRANSPARENT_STATEMENT_FILE \

>       --leaf $(jq -r .leaf <<<"$RESPONSE")"-foo"

Author: The leaf needs to be a valid hex value of 0x followed by 64 hex characters. In total representing 32 byte

Author: You could just link to the test case that does this with real event data

https://github.com/datatrails/datatrails-scitt-samples/blob/main/tests/test_verify_receipt.py#L52

as a developer I would find that example more useful

    ```

    The verification should fail with:

    ```output

>     ERROR:verify-receipt:failed to parse leaf hash

Author: The output for a failed verification is just "verification failed". ERROR output means the command line paramater was not accepted

1. Query DataTrails for the collection of statements

    ```bash
    PARAMS="event_attributes.subject=${SUBJECT}&page_size=1"
    curl "https://app.datatrails.ai/archivist/v2/publicassets/-/events?${PARAMS}" \
      | jq
    ```

    The events are listed starting with the most recently added.

{{< note >}}

> Coming soon: Filter on specific values, conveyed in the protected header. For example content types, such as what SBOMs have been registered, which issuers have made statements, or custom key value pairs.

Reviewer 2: ```suggestion
Coming soon: Filter on specific values conveyed in the protected header.
For example, content types, such as what SBOMs have been registered, which issuers have made statements or custom key-value pairs.
```

1. Query DataTrails for the collection of statements

    ```bash

>     PARAMS="event_attributes.subject=${SUBJECT}&page_size=1"

Reviewer 2: When running the samples multiple times, a user could wind up with a lot of statements for the subject "my-product-id".  For this reason, I added page_size to 3 to simulate a collection, but not overwhelm the screen with a long list that scrolls past their previous commands.

This PR changes it to 1. Was this intentional?

Reviewer 2: changing back to 3
```suggestion
    PARAMS="event_attributes.subject=${SUBJECT}&page_size=3"
```
